### PR TITLE
[Hub Menu] Navigation from MainTab using code is fixed

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Part of: #13693

## Description

To fix a Menu navigation crash that happens in iOS 18 beta we started using `TabContainerController` for the HubMenu tab. https://github.com/woocommerce/woocommerce-ios/pull/13671

This PR updates how we get the `HubMenuViewController` from the `MainTabBarController`. 

#### Coupons

- Log in to a store without any coupon
- Create an order from the app with a random product and select + Coupon.
- The empty coupon list presented would have a CTA to open the coupon screen.
- Tapping on the CTA should present an alert and then take you to the Coupons screen.

#### Payments

I found [this PR](https://github.com/woocommerce/woocommerce-ios/pull/11343) to test the payments navigation from Spotlight. However, the Woo payments options don't seem to be listed when searching from Spotlight. 

@joshheald Could you kindly help us test this part of the navigation?  

#### Privacy settings

- In `PrivacyBannerPresentationUseCase`, update `shouldShowPrivacyBanner` to return `true`.
- Build and launch the app
- Tap on "Go to Settings" from the privacy banner.
- Confirm that you are navigated to the privacy settings screen.

## Testing information

- I tested this in an iPhone running iOS 18 beta and an iOS simulator running iOS 17.4
- I didn't add unit tests because adding unit tests for the navigation wasn't straightforward. 
    - I could only test until the HubMenu, and I couldn't write tests for further navigation. 
    - Any suggestions are welcome. 

## Screenshots

| Privacy settings | Coupons |
|--------|--------|
| Simulator Screen Recording - iPhone 15 Pro Max - 2024-08-22 at 17 14 15 | Simulator Screen Recording - iPhone 15 Pro Max - 2024-08-22 at 17 15 08 | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/219">20.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift [hunk 1]
```diff
@@ -485,7 +489,9 @@ extension MainTabBarController {
     ///
     static func navigateToPrivacySettings() {
         switchToHubMenuTab {
-            guard let hubMenuViewController: HubMenuViewController = childViewController() else {
+            guard let hubMenuTabController = childViewController() as? TabContainerController,
+                  let navigationController = hubMenuTabController.wrappedController as? UINavigationController,
+                  let hubMenuViewController = navigationController.topViewController as? HubMenuViewController else {
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: Perhaps we could make `switchToHubMenuTab` pass the `hubMenuViewController` to its completion handler and avoid this duplication/make it easier to get right in future?

**Review by joshheald — COMMENTED** [review]
[Here's the code which should surface `Payments` to the spotlight search system](https://github.com/woocommerce/woocommerce-ios/blob/c708bb29626e1f613807a246814ea7e6b69a8b16/WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person%20Payments/Payments%20Menu/InPersonPaymentsMenuViewModel.swift#L120-L124).

It works by making a dummy view controller and setting it as the current context, with various keys needed to control the presentation. This is because we had loads of issues getting the SwiftUI equivalent to work, I think it needed a newer version of iOS as our minimum target.

Short story: it definitely won't show up in spotlight search until you first go to the payments menu.

I tested on iOS 17, by:

1. Doing a spotlight search for `Payments` – I could see it in the results
2. Deleting the app
3. Doing a spotlight search for `Payments`
4. Verifying that it didn't show up in the results
5. Launched the app, and logged in
6. Verified that a spotlight search still didn't show `Payments`
7. Navigated to `Menu > Payments`
8. Did a final spotlight search, and then I could see it in the results.

I don't have an iOS 18 device to test on but I think this is fine, no regression.

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: Good suggestion. Done in https://github.com/woocommerce/woocommerce-ios/pull/13697/commits/f7768039ea2b6906453fe2af23fad9f19cc1f993

**Comment by selanthiraiyan** [comment]
Thanks for testing the "Payments" case and reviewing the code, Josh!

@itsmeichigo @hafizrahman Could you kindly review the PR and test the "Coupons" and "Privacy settings" cases? Thank you!

**Review by itsmeichigo — APPROVED** [review]
I tested on simulators iOS 17.4 and 18.1 and confirmed that navigating to privacy settings works as expected. For coupons, around 30% of the time I got navigated to the Menu tab without seeing the Coupon list, but it's not consistent and not reproducible with the debugger on. I'm approving the PR disregard of this.

**Comment by selanthiraiyan** [comment]
Thanks for the review and testing, @itsmeichigo!

> For coupons, around 30% of the time I got navigated to the Menu tab without seeing the Coupon list, but it's not consistent and not reproducible with the debugger on. I'm approving the PR disregard of this.

I looked into this and added a comment [here](https://github.com/woocommerce/woocommerce-ios/issues/13693#issuecomment-2306117508). We can look into this later. 

I am merging this PR now to fix the other navigation cases.

**Pull request merged**